components/imageproc/src/lib.rs

```diff
@@ -0,0 +1,326 @@
+#[macro_use]
+extern crate lazy_static;
+extern crate regex;
+extern crate image;
+extern crate rayon;
+extern crate twox_hash;
+
+extern crate utils;
+extern crate errors;
+
+use std::path::{Path, PathBuf};
+use std::hash::{Hash, Hasher};
+use std::collections::HashMap;
+use std::collections::hash_map::Entry as HEntry;
+use std::fs::{self, File};
+
+use regex::Regex;
+use image::{GenericImage, FilterType};
+use image::jpeg::JPEGEncoder;
+use rayon::prelude::*;
+use twox_hash::XxHash;
+
+use utils::fs as ufs;
+use errors::{Result, ResultExt};
+
+
+static RESIZED_SUBDIR: &'static str = "_resized_images";
+lazy_static!{
+    pub static ref RESIZED_FILENAME: Regex = Regex::new(r#"([0-9a-f]{16})([0-9a-f]{2})[.]jpg"#).unwrap();
+}
+
+/// Describes the precise kind of a resize operation
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ResizeOp {
+    /// A simple scale operation that doesn't take aspect ratio into account
+    Scale(u32, u32),
+    /// Scales the image to a specified width with height computed such that aspect ratio is preserved
+    FitWidth(u32),
+    /// Scales the image to a specified height with width computed such that aspect ratio is preserved
+    FitHeight(u32),
+    /// Scales the image such that it fits within the specified width and height preserving aspect ratio.
+    /// Either dimension may end up being smaller, but never larger than specified.
+    Fit(u32, u32),
+    /// Scales the image such that it fills the specified width and height. Output will always have the exact dimensions specified.
+    /// The part of the image that doesn't fit in the thumbnail due to differing aspect ratio will be cropped away, if any.
+    Fill(u32, u32),
+}
+
+impl ResizeOp {
+    pub fn from_args(op: &str, width: Option<u32>, height: Option<u32>) -> Result<ResizeOp> {
+        use ResizeOp::*;
+
+        // Validate args:
+        match op {
+            "fitwidth" => if width.is_none() { return Err(format!("op=fitwidth requires a `width` argument").into()) },
```

Comment:
can we have `fit_width` and `fit_height` to match the usual type of arguments?

Reply:
Yes, there was also a bug on the next line (`fitwidth` was copypasted over).